Fix GPU card validator

samaradel opened the pull request:
### Description

Validate the GPU card selector before deployment
 
### Changes

[Screencast from 24-06-25 15:51:26.webm](https://github.com/user-attachments/assets/f2607829-c9c5-4bb6-9bc3-432a418eedc5)


### Related Issues

- https://github.com/threefoldtech/tfgrid-sdk-ts/issues/4146

### Tested Scenarios

- Select node with the GPU and deploy without selecting the GPU card 

### Documentation PR

For UI changes, Please provide the Documentation PR on [info_grid](https://github.com/threefoldtech/info_grid)

### To consider

#### Preliminary Checks:

- [x] Preliminary Checks
  - Does it completely address the issue linked?
  - What about edge cases?
  - Does it meet the specified acceptance criteria?
  - Are there any unintended side effects?
  - Does the PR adhere to the team's coding conventions, style guides, and best practices?
  - Does it integrate well with existing features?
  - Does it impact the overall performance of the application?
  - Are there any bottlenecks or slowdowns?
  - Has it been optimized for efficiency?
  - Has it been adequately tested with unit, integration, and end-to-end tests?
  - Are there any known defects or issues?
  - Is the code well-documented?
  - Are changes to documentation reflected in the code?

#### UI Checks:

- [x] UI Checks
  - If a UI design is provided/ does it follow it?
  - Does every button work?
  - Is the data displayed logical? Is it what you expected?
  - Does every validation work?
  - Does this interface feel intuitive?
  - What about slow network? Offline?
  - What if the gridproxy/graphql/chain is failing?
  - What would a first time user have a hard time navigating here?

#### Code Quality Checks:

- [x] Code Quality Checks
  - Code formatted/linted? Are there unused imports? .. etc
  - Is there redundant/repeated code?
  - Are there conditionals that are always true or always false?
  - Can we write this more concisely?
  - Can we reuse this code? If yes, where?
  - Will the changes be easy to maintain and update in the future?
  - Can this code become too complex to understand for other devs?
  - Can this code cause future integration problems?

### Testing Checklist

- [ ] Does it Meet the specified acceptance criteria.
- [ ] Test if changes can affect any other functionality.
- [ ] Tested with unit, integration, and end-to-end tests.
- [ ] Tested the un-happy path and rollback scenarios.
- [ ] Documentation updated to meet the latest changes.
- [ ] Check automated tests working successfully with the changes.
- [ ] Can be covered by automated tests.

### General Checklist

- [ ] Tests included
- [x] Build pass
- [ ] Documentation
- [x] Code format and docstring
- [x] Screenshots/Video attached (needed for UI changes)

Commit:
Fix GPU card validator

0oM4R commented:
Can you please explain the solution?


 I think we should fix that on the formValidator, as this is happening on the domain name as well 
 
[Screencast from 2025-06-29 10-51-11.webm](https://github.com/user-attachments/assets/65cd9d4e-6852-4c8c-b8b3-bbc7cad7eaae)

samaradel commented:
well, 
- The class `gpu-card-selector` was added as styling for the GPU selector container, `data-validation-target="gpu-cards"` that serves as a marker for the validation and scrolling system
- By adding `ref="gpuSelect"`, we can directly access the VSelect component instance
- Calling `bindStatus()` without forcing a specific status. Instead, it uses the improved logic inside the bindStatus function
- `targetElement` used as a computed property so it can get the VSelect's DOM Element to use it with `$el`
- `registrationId` initialized as an id for the component, and `isRegistered` to track if it's  registered with the form
- `registerService` to register if it's not already registered and to update its `isRegistered` flag
- `unregisterService` same service as the last point, but for unregister
- The `onMounted` is updated to only register if there's a valid node selected
- Update the deep watcher with the `registerService` and `unregisterService`

P.S. I made the changes in the `Select GPU` component itself, if it's exists for the domain component too, please open an issue for it

amiraabouhadid commented on packages/playground/src/components/node_selector/TfSelectGpu.vue:
```diff
@@ -82,39 +84,109 @@ export default {
 
     function bindStatus(status?: ValidatorStatus) {
-      const s = status || ValidatorStatus.Init;
+      let s = status;
+
+      if (!s) {
+        if (props.validNode && props.node && props.modelValue.length === 0) {
+          s = ValidatorStatus.Invalid;
+        } else if (props.validNode && props.node && props.modelValue.length > 0) {
+          s = ValidatorStatus.Valid;
+        } else {
+          s = ValidatorStatus.Init;
+        }
+      }
+
```
```suggestion
let s = status;

if (!s) {
  if (props.validNode && props.node) {
    s = props.modelValue.length === 0
      ? ValidatorStatus.Invalid
      : ValidatorStatus.Valid;
  } else {
    s = ValidatorStatus.Init;
  }
}

```

amiraabouhadid commented:
correctly highlights unselected gpu cards 
<img width="848" alt="Screenshot 2025-06-29 at 21 30 00" src="https://github.com/user-attachments/assets/a416de60-001c-4267-a053-8b99826c09f6" />

amiraabouhadid commented:
correctly highlights unselected gpu cards 
<img width="848" alt="Screenshot 2025-06-29 at 21 30 00" src="https://github.com/user-attachments/assets/a416de60-001c-4267-a053-8b99826c09f6" />

Commit:
Refactor and reduce checking node validation

amiraabouhadid commented:
waiting on mainnet tokens to continue testing. also encountered this https://github.com/threefoldtech/tfgrid-sdk-ts/issues/4261#issuecomment-3024052870

0oM4R commented:
> P.S. I made the changes in the `Select GPU` component itself, if it's exists for the domain component too, please open an issue for it

@samaradel 

What I'm talking about is checking if the fix can be done on the parent component, so that will solve both issues.

amiraabouhadid reviewed (dismissed):
works 
<img width="943" height="261" alt="Screenshot 2025-07-13 at 14 18 16" src="https://github.com/user-attachments/assets/857a5a92-d548-4d16-8c1b-89deea5de071" />
please address @0oM4R comment before merging

0oM4R commented:
> > P.S. I made the changes in the `Select GPU` component itself, if it's exists for the domain component too, please open an issue for it
> 
> @samaradel
> 
> What I'm talking about is checking if the fix can be done on the parent component, so that will solve both issues.



This is not a blocking request. If you are sure that the fix is only related to the GPU component, go ahead and merge, and we can handle the domain name in another issue

Commit:
- Create form validation helpers file for highlighting and scrolling
- Add data dttribute targeting
- Add custom CSS selector
- Add same error highlighting across all components

Commit:
Merge branch 'development' into dev_GPU_card

Commit:
Remove unused value to fix linting

samaradel commented:
@0oM4R @amiraabouhadid, it's all changed now as per https://github.com/threefoldtech/tfgrid-sdk-ts/pull/4253#issuecomment-3067036622 please review

amiraabouhadid commented:
please resolve conflicts

Commit:
Merge branch 'development' into dev_GPU_card

amiraabouhadid reviewed (commented):
no available gpu nodes on devnet to test with, waiting on funds on mainnet to test

amiraabouhadid reviewed (commented):
works 
<img width="1203" height="649" alt="Screenshot 2025-07-21 at 15 10 09" src="https://github.com/user-attachments/assets/fd91ac4c-153b-4a40-8765-5944be8242dc" />
